### ISSUE TITLE
Separate to Ref::Util::XS

### ISSUE THREAD

**Issue by xsawyerx** [p5pclub/ref-util#26]
`Ref::Util` should have a dynamic config that allows adding `Ref::Util::XS` to the dependencies if there's a compiler. It's less than ideal but it's the best we can do. `Ref::Util` should include the PP versions, available in `Ref::Util::MaybeXS` (not released).

There's also the trick of `Devel::GlobalDestruction` which should be taken into account when a dist is moved to an optional XS setup.

**Comment by xsawyerx** [comment]
I started work on this but unfortunately needed to let go so we could release the next version with fixes for 5.8 and 5.6. See #34.

**Comment by iynehz** [comment]
We really need a PP version. I find this module cannot build with MSVC or Solaris Studio 12.

**Comment by arc** [comment]
@stphnlyd There will definitely be a PP version at some point for the benefit of people with no compiler, but it would also be good to fix those build failures. I can't see any smoke-test failures in http://matrix.cpantesters.org/?dist=Ref-Util+0.111 but maybe I'm looking in the wrong place or something. If you have any more information about the failures under those compilers, that would be really helpful — thanks!

**Comment by iynehz** [comment]
I just made a PR for some versions of MSVC and Sun Studio. https://github.com/p5pclub/ref-util/pull/35

The problem with some older versions of MSVC is about C90-compatibility. 

The Sun Studio fix is about that last ternary operator, seems that expansion of the macros could result some complex expresions that do not work within the ternary operation. For example, the error on Perl 5.20, Sun Studio 12.1 was like below. I even tried adding parentheses but it did not work. Changing it to if-else helped. 

```
"Util.xs", line 246: syntax error before or at: do
"Util.xs", line 246: syntax error before or at: :
"Util.xs", line 246: cannot recover from previous errors
cc: acomp failed for Util.c
```

**Comment by arc** [comment]
Ref::Util::XS has been released, and Ref::Util 0.200 uses it whenever possible, falling back to Ref::Util::PP otherwise.
